src/dbnode/storage/index/block.go

```diff
@@ -304,19 +304,26 @@ func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
 		return b.writeBatchResult(inserts, MutableSegmentsStats{},
 			b.writeBatchErrorInvalidState(b.state))
 	}
-	if b.state == blockStateSealed {
-		coldBlock := b.coldMutableSegments[len(b.coldMutableSegments)-1]
-		b.RUnlock()
-		_, err := coldBlock.WriteBatch(inserts)
-		// Don't pass stats back from insertion into a cold block,
-		// we only care about warm mutable segments stats.
-		return b.writeBatchResult(inserts, MutableSegmentsStats{}, err)
-	}
 	b.RUnlock()
 	stats, err := b.mutableSegments.WriteBatch(inserts)
 	return b.writeBatchResult(inserts, stats, err)
 }
 
+func (b *block) WriteColdBatch(inserts *WriteBatch) (WriteBatchResult, error) {
+	b.RLock()
+	if !b.writesAcceptedWithRLock() {
+		b.RUnlock()
+		return b.writeBatchResult(inserts, MutableSegmentsStats{},
+			b.writeBatchErrorInvalidState(b.state))
+	}
+	coldBlock := b.coldMutableSegments[len(b.coldMutableSegments)-1]
+	b.RUnlock()
+	_, err := coldBlock.WriteBatch(inserts)
+	// Don't pass stats back from insertion into a cold block,
+	// we only care about warm mutable segments stats.
```

Comment:
Hm, can't we make that decision in the caller not to use the stats?  Or does the "Combine()" method do it automatically? Could we make it a "CombineWithoutStats()" so that at least this doesn't respond with just an empty datastructure that we don't really know should be empty or not from the viewpoint of the caller?